src/main/java/com/google/sps/servlets/GenerateServlet.java

```diff
@@ -26,16 +30,116 @@
 @WebServlet("/generate-hunt")
 public class GenerateServlet extends HttpServlet {
 
-  private static final String FILTER_ARRAY = "clicked-array";
+  private static final String PLACE_FILTERS = "user-places";
+  private static final String DIFF_FILTERS = "user-diff";
+
+  String[] allPlaces = {"Paris", "New York City", "San Francisco", "London", "Sydney", "Venice"};
+  String[] allDifficulties = {"Easy", "Medium", "Hard"};
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Get array of clicked filters and convert to ArrayList<String>
-    Gson gson = new Gson();
-    HashSet<String> clickedFilters =
-        gson.fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
+
+    HashSet<String> userPlaces =
+        new Gson().fromJson(request.getParameter(PLACE_FILTERS), HashSet.class);
+    HashSet<String> userDifficultyStrings =
+        new Gson().fromJson(request.getParameter(DIFF_FILTERS), HashSet.class);
+
+    // Create fake destinations TODO: get destinations from datastore
+    ArrayList<Destination> allDestinations = createFakeDestinations();
+
+    // Convert difficulty level strings to Destination.Obscurity
+    HashSet<Destination.Obscurity> userDifficultyLevels = new HashSet();
+    for (String level : userDifficultyStrings) {
+      userDifficultyLevels.add(Destination.stringToEnum(level));
+    }
+    if (userDifficultyStrings.size() == 0) {
+      userDifficultyLevels.add(Destination.stringToEnum("Easy"));
+      userDifficultyLevels.add(Destination.stringToEnum("Medium"));
+      userDifficultyLevels.add(Destination.stringToEnum("Hard"));
+    }
+
+    // Filter
+    Set<Destination> filteredDestinations =
+        filter(allDestinations, userPlaces, userDifficultyLevels);
+
+    writeToDataStore(filteredDestinations);
 
     response.setContentType("text/html;");
-    response.getWriter().println(clickedFilters);
+    response.getWriter().println(filteredDestinations);
+  }
+
+  /* Return ArrayList<Destination> of filtered Destination objects. */
+  public Set<Destination> filter(
```

Comment:
Now that we're adding more complicated filtering logic here, I think you should write tests to make sure your filter function is working as intended!  Can you add a test file with a couple tests to demonstrate that it returns the expected Destinations?